### PR TITLE
feat(Free Web Novel): New presence

### LINKED ISSUE #6035: Free Web Novel | freewebnovel.com
### Website name

Free Web Novel

### Website URL

https://freewebnovel.com/

### Website logo

https://files.catbox.moe/acvgh1.png

### Prerequisites

- [ ] It is a paid service
- [ ] It displays NSFW content
- [ ] It is region restricted

### Description

Adding [Free Web Novel](https://freewebnovel.com/) as a pressence where when you are on the site it shows cover, browsing and button to it

When you are on a novel it shows cover of it on the pressence and button for it to the site
When you are on a chapter it shows cover for novel and buttons for it to the chapter and novel on site

Playing game: Free Web Novel
Logo to site when browsing 
Browsing Novel's 
00:01:53 elapsed
Button 1: Visit Site (link to site)

Playing game: Free Web Novel
Cover to novel you are on
Novel Title
00:11:48 elapsed
Button 1: Novel Title (links to novel)

Playing game: Free Web Novel
Cover to novel you are on
Novel Title
Chapter 123: wandering coder
00:03:13 elapsed
Button 1: Novel Title (links to novel)
Button 2: Chapter 123: Wandering Coder (links to chapter)

Alternative logo 400x400: (quick mockup of their site logo so it would fit better with discord cover for pressence)
with background  https://de.catbox.moe/5spncb.png
bilde

transparent background https://de.catbox.moe/nrx1xs.png
bilde

### PR DESCRIPTION
**Pull request by BigBoi-TM**
**Describe the changes in this pull request:**
#
# (Feature) New presence for Free Web Novel.
Resolves #6035  
Features:
- Detects pages from the novel list along with pages for different genres
- Detects a when a novel is being viewed and/or read
    -  Displays the novel being viewed and the chapter when being read
- Includes multiple settings to go incognito, and to disable/enable certain things to display
#
<summary> Proof showing the creation/modification is working as expected </summary>
<br>

<details>
<summary>Screenshots</summary>

### - Homepage
image

### - Author Page
image

### - Page Detection
image

### - Reading a Novel
image

### - Reading Detection
image
image

### - Settings
image

</details>

### CODE REVIEW

File: websites/F/Free Web Novel/presence.ts [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+		showLogo = true;
+		showButtons = false;
+	}
+	if (!showBook) showCover = false;
+
+	if (showButtons === true) {
+		presenceData.buttons = [
+			{ label: "Visit freewebnovel.com", url: "https://freewebnovel.com/" }
+		];
+	} else presenceData.buttons = null;
```
Comment (Slowlife01) [inline websites/F/Free Web Novel/presence.ts:53]: No buttons on homepage
```suggestion
```

File: websites/F/Free Web Novel/presence.ts [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+	switch (true) {
+		case privacy: {
+			showButtons = false;
+			presenceData.state = null;
```
Comment (Slowlife01) [inline websites/F/Free Web Novel/presence.ts:57]:
```suggestion
```
Unnecessary

File: websites/F/Free Web Novel/presence.ts [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+				pathnames.includes(pathname) ||
+				document.querySelector<HTMLDivElement>('[class="m-imgtxt"]')
+			)
+				presenceData.details = showReading ? "Browsing..." : null;
+			else if (
+				document.querySelector<HTMLAnchorElement>('[title="Read Next chapter"]')
+			) {
+				presenceData.details = showReading ? "Reading..." : null;
+				presenceData.smallImageKey = showReading ? "opem" : null;
+			}
```
Comment (Slowlife01) [inline websites/F/Free Web Novel/presence.ts:69]: Just add showReading into the if statement

File: websites/F/Free Web Novel/presence.ts [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+		}
+		case showLogo: {
+			showTimestamp = false;
+			showButtons = false;
+			showCover = false;
+			showBook = false;
+			presenceData.largeImageKey = "fwn_1024";
+			presenceData.smallImageKey = null;
+			presenceData.details = null;
+			presenceData.state = null;
```
Comment (Slowlife01) [inline websites/F/Free Web Novel/presence.ts:80]:
```suggestion
			presenceData.largeImageKey = "fwn_1024";
			delete presenceData.smallImageKey;
```

**Comment by BigBoi-TM** [comment]
Note: May or may not contain nsfw. Dunno, bunch of suggestive stuff on there

File: websites/F/Free Web Novel/dist/metadata.json [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+	"thumbnail": "https://i.imgur.com/PJw9FKX.png",
+	"color": "#5a5a5a",
+	"category": "socials",
+	"tags": [
+	  "novel",
+	  "book",
+	  "books",
+	  "reading",
+	  "free"
+	],
```
Comment (EncryptedDev) [inline websites/F/Free Web Novel/dist/metadata.json:23]: Mix of tabs an sapces, format the file properly using CLI

**Review by EncryptedDev — CHANGES_REQUESTED** [review]

**Comment by EncryptedDev** [comment]
Resolve Slowlife's requested changes once handled please

**Comment by BigBoi-TM** [comment]
> Resolve Slowlife's requested changes once handled please

I got no reply on two requests, do I still have to change them?

**Comment by BigBoi-TM** [comment]
@EncryptedDev I miss anything?

File: websites/F/Free Web Novel/presence.ts (hunk 3 above)
Comment (Slowlife01) [reply websites/F/Free Web Novel/presence.ts:61]:
```ts
			if (document.querySelector('[title="Read Next chapter"]') && showReading) {
				presenceData.details = "Reading...";
				presenceData.smallImageKey = "opem";
			}
```

**Comment by BigBoi-TM** [comment]
Copied an old version, tryna fix it

File: websites/F/Free Web Novel/presence.ts [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+			break;
+		}
+		case pathname === "/search/": {
+			presenceData.details = "Searching for a novel";
+			presenceData.state = "Page: Search";
+			break;
+		}
+		case pathname === "/contact.html": {
+			presenceData.details = "Veiwing the contact page";
+			presenceData.state = "Page: Contact";
```
Comment (EncryptedDev) [inline websites/F/Free Web Novel/presence.ts:136]: This is a given as the details above says the same thing

**Review by EncryptedDev — CHANGES_REQUESTED** [review]
Quality code, just way too much detail for a presence really

File: websites/F/Free Web Novel/presence.ts [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+			"/completed-novel/",
+			"/most-popular-novel/",
+			"/search/",
+			"/contact.html",
+			"/sitemap.xml",
+			"/privacy-terms-of-use.html",
+			"/genre/",
+			"/author/"
+		];
+	if (showTimestamp) presenceData.startTimestamp = browsingTimestamp;
```
Comment (EncryptedDev) [inline websites/F/Free Web Novel/presence.ts:38]: If you are using a ternary in the presenceData object for the other values, you might as well include this too

File: websites/F/Free Web Novel/presence.ts [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+		}
+		case pathname.includes("/most-popular-novel/"): {
+			presenceData.details = "Browsing the most popular novels";
+			break;
+		}
+		case pathname.includes("/genre/"): {
+			const genre =
+				document.querySelector<HTMLHeadingElement>('[class="tit"]').textContent;
+			presenceData.details = `Checking out ${genre.toLowerCase()}`;
+			presenceData.state = `Genre: ${genre}`;
```
Comment (EncryptedDev) [inline websites/F/Free Web Novel/presence.ts:105]: This is repeated?

File: websites/F/Free Web Novel/presence.ts [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+			}
+			break;
+		}
+		case pathname === "/search/": {
+			presenceData.details = "Searching for a novel";
+			presenceData.state = "Page: Search";
+			break;
+		}
+		case pathname === "/contact.html": {
+			presenceData.details = "Veiwing the contact page";
```
[the remark this answers is not in the document]
Comment (EncryptedDev) [reply websites/F/Free Web Novel/presence.ts:128]: Unresolved

File: websites/F/Free Web Novel/presence.ts [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+			break;
+		}
+		case pathname === "/sitemap.xml": {
+			presenceData.details = "Viewing the site's sitemap";
+			break;
+		}
+		case pathname === "/privacy-terms-of-use.html": {
+			presenceData.details = "Looking through the privacy & terms of use";
```
Comment (EncryptedDev) [inline websites/F/Free Web Novel/presence.ts:136]: Very long

**Review by EncryptedDev — CHANGES_REQUESTED** [review]

File: websites/F/Free Web Novel/dist/metadata.json [hunk 11]
```diff
@@ -0,0 +1,96 @@
+{
+	"$schema": "https://schemas.premid.app/metadata/1.7",
+	"author": {
+		"name": "BigBoi©",
```
Comment (ImRodry) [inline websites/F/Free Web Novel/dist/metadata.json:4]: you're not copyrighted
```suggestion
		"name": "BigBoi",
```

File: websites/F/Free Web Novel/dist/metadata.json [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+		{
+			"id": "showButtons",
+			"if": {
+				"privacy": false,
+				"showLogo": false
+			},
+			"title": "Show Buttons",
+			"icon": "fas fa-compress-arrows-alt",
+			"value": true
+
```
Comment (ImRodry) [inline websites/F/Free Web Novel/dist/metadata.json:93]: please remove the extra empty lines in these settings

File: websites/F/Free Web Novel/dist/metadata.json [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+		},
+		{
+			"id": "showCover",
+			"if": {
+				"privacy": false,
+				"showLogo": false,
+				"showBook": true,
+				"showReading": true
```
Comment (ImRodry) [inline websites/F/Free Web Novel/dist/metadata.json:62]: were all of these tested to their full extent?

File: websites/F/Free Web Novel/presence.ts [hunk 14]
```diff
@@ -0,0 +1,185 @@
+const presence = new Presence({
+		clientId: "965294297048023050"
+	}),
+	browsingTimestamp = Math.floor(Date.now());
```
Comment (ImRodry) [inline websites/F/Free Web Novel/presence.ts:4]: what
```suggestion
	browsingTimestamp = Math.round(Date.now() / 1000);
```

File: websites/F/Free Web Novel/presence.ts [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+				presenceData.smallImageKey = "open";
+			}
+			break;
+		}
+		case showLogo: {
+			presenceData.largeImageKey = "fwn_1024";
+			delete presenceData.smallImageKey;
+			delete presenceData.details;
+			delete presenceData.state;
+			delete presenceData.buttons;
```
Comment (ImRodry) [inline websites/F/Free Web Novel/presence.ts:73]: this needs to be ran at the end of the file I believe

**Review by ImRodry — CHANGES_REQUESTED** [review]

**Comment by ImRodry** [comment]
@BigBoi-TM please don't resolve conversations without addressing them in some way

**Comment by BigBoi-TM** [comment]
> @BigBoi-TM please don't resolve conversations without addressing them in some way

Im waiting on a reply for two comments.

**Comment by ImRodry** [comment]
@BigBoi-TM I don't see any questions you asked, only some unresolved reviews

**Comment by BigBoi-TM** [comment]
@ImRodry 
image

**Comment by ImRodry** [comment]
It says pending. I can't see your comments if they're pending because you didn't send them...

**Comment by ImRodry** [comment]
Hey! Thank you for making this PR. Unfortunately, I will be closing this after multiple unsuccessful attempts at getting reviews addressed by you and a lack of communication in the last few days. Should you wish to continue working on this, please address the reviews and notify a member of the reviewer team.

**Comment by BigBoi-TM** [comment]
Bruh I've been sending comments left and right none of them want to go through 😭

**Pull request closed without merging**